polkadot/runtime/common/src/integration_tests.rs

```diff
@@ -579,6 +601,493 @@ fn basic_end_to_end_works() {
+		run_to_session(START_SESSION_INDEX);
+
+		// User 1 will own a parachain
+		let free_balance = 1_000_000_000;
+		Balances::make_free_balance_be(&account_id(1), free_balance);
+		// Register an on demand parachain
+		let mut code_size = 1024 * 1024;
+		let genesis_head = Registrar::worst_head_data();
+		let head_size = genesis_head.0.len();
+		let code_0 = validation_code(code_size);
+
+		assert_ok!(Registrar::reserve(signed(1)));
+		assert_ok!(Registrar::register(
+			signed(1),
+			ParaId::from(para_id),
+			genesis_head.clone(),
+			code_0.clone(),
+		));
+		conclude_pvf_checking::<Test>(&code_0, VALIDATORS, START_SESSION_INDEX, true);
+
+		// The para should be onboarding.
+		assert_eq!(Paras::lifecycle(ParaId::from(para_id)), Some(ParaLifecycle::Onboarding));
+		// After two sessions the parachain will be succesfully registered as an on-demand.
+		run_to_session(START_SESSION_INDEX + 2);
+		assert_eq!(Paras::lifecycle(ParaId::from(para_id)), Some(ParaLifecycle::Parathread));
+		// The deposit should be appropriately taken.
+		let total_bytes_stored = code_size as u32 + head_size as u32;
+		assert_eq!(
+			Balances::reserved_balance(&account_id(1)),
+			ParaDeposit::get() + (total_bytes_stored * DataDepositPerByte::get())
+		);
+
+		// We 'simulate' the migration. In this case there is only one legacy parachain.
+		paras_registrar::LegacyParas::<Test>::insert(ParaId::from(para_id), ());
+
+		// The parachain should be able to schedule a code upgrade without any additional deposit or
+		// fees necessary.
+
+		let location: MultiLocation = (Parent, Parachain(para_id.into())).into();
+		let sovereign_account =
+			<Test as paras_registrar::Config>::SovereignAccountOf::convert_location(&location)
+				.unwrap();
+		let para_origin: runtime_parachains::Origin = u32::from(LOWEST_PUBLIC_ID).into();
+
+		Balances::make_free_balance_be(&sovereign_account, free_balance);
+
+		code_size *= 2;
+		let code_1 = validation_code(code_size);
+		assert_ok!(Registrar::schedule_code_upgrade(
+			para_origin.clone().into(),
+			ParaId::from(para_id),
+			code_1.clone(),
+		));
+		conclude_pvf_checking::<Test>(&code_1, VALIDATORS, START_SESSION_INDEX + 2, true);
+
+		// After two more sessions the parachain can be upgraded.
+		run_to_session(START_SESSION_INDEX + 4);
+		// Force a new head to enact the code upgrade.
+		assert_ok!(Paras::force_note_new_head(
+			RuntimeOrigin::root(),
+			para_id,
+			genesis_head.clone()
+		));
+		assert_eq!(Paras::current_code(&para_id), Some(code_1.clone()));
+
+		// Should remain unchanged:
+		assert_eq!(
+			Balances::reserved_balance(&account_id(1)),
+			ParaDeposit::get() + (total_bytes_stored * DataDepositPerByte::get())
+		);
+		assert_eq!(Balances::total_balance(&account_id(1)), free_balance);
+		assert_eq!(Balances::total_balance(&sovereign_account), free_balance);
+
+		// However, legacy chains are only allowed to perform a free upgrade once.
+		// Doing a new upgrade will require fees:
```

Comment:
My idea was to allow this free upgrade so that any ongoing upgrades initiated before incurring extra costs would proceed. Initially, my thought process was that, afterward, parachains should be made aware of the change and adjust before proceeding with additional code upgrades. However, I now realize that this requirement for parachains to prepare is too rigorous.

I will implement the solution you described below since it is much more forgiving to parachains and less complex.